internal/query/rolebindings.go

```diff
@@ -0,0 +1,624 @@
+			OptionalSubjectId: rb.ID.String(),
+		},
+	})
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	// create a list of delete updates for these relationships
+	updates := make([]*pb.RelationshipUpdate, len(fromRels)+len(toRels))
+
+	for i, rel := range append(fromRels, toRels...) {
+		updates[i] = &pb.RelationshipUpdate{
+			Operation:    pb.RelationshipUpdate_OPERATION_DELETE,
+			Relationship: rel,
+		}
+	}
+
+	// apply changes
+	if _, err := e.client.WriteRelationships(ctx, &pb.WriteRelationshipsRequest{Updates: updates}); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	return nil
+}
+
+func (e *engine) ListRoleBindings(ctx context.Context, resource types.Resource, optionalRole *types.Resource) ([]types.RoleBinding, error) {
+	ctx, span := e.tracer.Start(
+		ctx, "engine.ListRoleBinding",
+		trace.WithAttributes(
+			attribute.Stringer("resource_id", resource.ID),
+		),
+	)
+	defer span.End()
+
+	e.logger.Debugf("listing role-bindings for resource: %s, optionalRole: %v", resource.ID, optionalRole)
+
+	// 1. list all grants on the resource
+	listRbFilter := &pb.RelationshipFilter{
+		ResourceType:       e.namespaced(resource.Type),
+		OptionalResourceId: resource.ID.String(),
+		OptionalRelation:   iapl.GrantRelationship,
+		OptionalSubjectFilter: &pb.SubjectFilter{
+			SubjectType: e.namespaced(e.rbac.RoleBindingResource.Name),
+		},
+	}
+
+	grantRel, err := e.readRelationships(ctx, listRbFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. fetch role-binding details for each grant
+	bindings := make([]types.RoleBinding, 0, len(grantRel))
+	errs := make([]error, 0, len(grantRel))
+
+	for _, rel := range grantRel {
+		rbRes, err := e.NewResourceFromIDString(rel.Subject.Object.ObjectId)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		rb, err := e.GetRoleBinding(ctx, rbRes)
+		if err != nil {
+			if errors.Is(err, ErrRoleBindingNotFound) {
+				// print and record a warning message when there's a grant points
+				// to a role-binding that not longer exists.
+				//
+				// this should not happen in normal circumstances, but it's possible
+				// if some role-binding relationships are deleted directly through
+				// spiceDB
+				err := fmt.Errorf("%w: dangling grant relationship: %s", err, rel.String())
+
+				span.RecordError(err)
+				e.logger.Warnf(err.Error())
+			}
+
+			errs = append(errs, err)
+
+			continue
+		}
+
+		if optionalRole != nil && rb.Role.ID.String() != optionalRole.ID.String() {
+			continue
+		}
+
+		if len(rb.Subjects) == 0 {
+			continue
+		}
+
+		bindings = append(bindings, rb)
+	}
+
+	for _, err := range errs {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return bindings, nil
+}
+
+func (e *engine) UpdateRoleBinding(ctx context.Context, rb types.Resource, subjects []types.RoleBindingSubject) (types.RoleBinding, error) {
```

Comment:
Something we need to think about here is locking to ensure we serialize changes to a role binding, or even a resource. As written, we're set up for a number of potential race condition problems for create, update, and delete of bindings.